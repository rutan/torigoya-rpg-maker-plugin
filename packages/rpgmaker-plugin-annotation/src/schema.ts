export {
  StringOrI18nString,
  StructOrI18nStruct,
  StructArrayOrI18nStructArray,
  PluginParameterBase,
  PluginParameterString,
  PluginParameterStringArray,
  PluginParameterMultilineString,
  PluginParameterMultilineStringArray,
  PluginParameterNote,
  PluginParameterNoteArray,
  PluginParameterFile,
  PluginParameterFileArray,
  PluginParameterNumber,
  PluginParameterNumberArray,
  PluginParameterBoolean,
  PluginParameterBooleanArray,
  PluginParameterSelect,
  PluginParameterSelectArray,
  PluginParameterCombo,
  PluginParameterDatabase,
  PluginParameterDatabaseArray,
  PluginParameterStruct,
  PluginParameterStructArray,
  PluginParameter,
  PluginStruct,
  PluginCommand,
  PluginNote,
  PluginConfigSchema,
} from './internal/schemaDefine.js';
