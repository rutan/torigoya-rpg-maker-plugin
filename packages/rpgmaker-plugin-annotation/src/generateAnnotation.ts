import { PluginConfigSchema, PluginParameter } from './schema.js';

export interface BuildOptions {
  languages: string[];
  defaultLanguage: string;
}

// ISO 639 language codes
const localeKeyISO639_1 = [
  'aa',
  'ab',
  'ae',
  'af',
  'ak',
  'am',
  'an',
  'ar',
  'as',
  'av',
  'ay',
  'az',
  'ba',
  'be',
  'bg',
  'bh',
  'bi',
  'bm',
  'bn',
  'bo',
  'br',
  'bs',
  'ca',
  'ce',
  'ch',
  'co',
  'cr',
  'cs',
  'cu',
  'cv',
  'cy',
  'da',
  'de',
  'dv',
  'dz',
  'ee',
  'el',
  'en',
  'eo',
  'es',
  'et',
  'eu',
  'fa',
  'ff',
  'fi',
  'fj',
  'fo',
  'fr',
  'fy',
  'ga',
  'gd',
  'gl',
  'gn',
  'gu',
  'gv',
  'ha',
  'he',
  'hi',
  'ho',
  'hr',
  'ht',
  'hu',
  'hy',
  'hz',
  'ia',
  'id',
  'ie',
  'ig',
  'ii',
  'ik',
  'io',
  'is',
  'it',
  'iu',
  'ja',
  'jv',
  'ka',
  'kg',
  'ki',
  'kj',
  'kk',
  'kl',
  'km',
  'kn',
  'ko',
  'kr',
  'ks',
  'ku',
  'kv',
  'kw',
  'ky',
  'la',
  'lb',
  'lg',
  'li',
  'ln',
  'lo',
  'lt',
  'lu',
  'lv',
  'mg',
  'mh',
  'mi',
  'mk',
  'ml',
  'mn',
  'mr',
  'ms',
  'mt',
  'my',
  'na',
  'nb',
  'nd',
  'ne',
  'ng',
  'nl',
  'nn',
  'no',
  'nr',
  'nv',
  'ny',
  'oc',
  'oj',
  'om',
  'or',
  'os',
  'pa',
  'pi',
  'pl',
  'ps',
  'pt',
  'qu',
  'rm',
  'rn',
  'ro',
  'ru',
  'rw',
  'sa',
  'sc',
  'sd',
  'se',
  'sg',
  'si',
  'sk',
  'sl',
  'sm',
  'sn',
  'so',
  'sq',
  'sr',
  'ss',
  'st',
  'su',
  'sv',
  'sw',
  'ta',
  'te',
  'tg',
  'th',
  'ti',
  'tk',
  'tl',
  'tn',
  'to',
  'tr',
  'ts',
  'tt',
  'tw',
  'ty',
  'ug',
  'uk',
  'ur',
  'uz',
  've',
  'vi',
  'vo',
  'wa',
  'wo',
  'xh',
  'yi',
  'yo',
  'za',
  'zh',
  'zu',
];

export function generateAnnotation(config: PluginConfigSchema, { languages, defaultLanguage }: BuildOptions) {
  return languages
    .map((language) => {
      const omitLanguageSignature = language === defaultLanguage;

      function pickString(value: string | { [key: string]: string }) {
        if (typeof value === 'string') return value;
        if (Object.prototype.hasOwnProperty.call(value, defaultLanguage)) return value[language];
        if (Object.prototype.hasOwnProperty.call(value, defaultLanguage)) return value[defaultLanguage];
        return '';
      }

      function pickObject<T extends { [key: string]: any }>(value: T | { [key: string]: T }) {
        const keys = Object.keys(value);
        if (keys.length > 0 && keys.every((key) => localeKeyISO639_1.includes(key))) {
          return value[language] || value[defaultLanguage];
        } else {
          return value;
        }
      }

      function push(lines: string[], annotation: string, stringValue: string) {
        stringValue.split(/\r?\n/).forEach((line, i) => {
          lines.push(`${i === 0 ? `@${annotation} ` : ''}${line}`);
        });
      }

      function pushParameter(lines: string[], name: string, param: PluginParameter) {
        push(lines, name, param.name);

        if (param.text) push(lines, 'text', pickString(param.text));
        if (param.description) push(lines, 'desc', pickString(param.description));
        if (param.parent) push(lines, 'parent', pickString(param.parent));

        // type
        switch (param.type) {
          case 'struct':
            push(lines, 'type', `struct<${param.struct}>`);
            break;
          case 'struct[]':
            push(lines, 'type', `struct<${param.struct}>[]`);
            break;
          default:
            if (config.target.includes('MV') && param.type.startsWith('multiline_string')) {
              push(lines, 'type', param.type.replace('multiline_string', 'note'));
            } else {
              push(lines, 'type', param.type);
            }
        }

        switch (param.type) {
          case 'string':
            if (param.default) push(lines, 'default', pickString(param.default));
            break;
          case 'string[]':
            push(lines, 'default', JSON.stringify(param.default.map((n) => pickString(n))));
            break;
          case 'multiline_string':
            if (param.default) {
              if (config.target.includes('MV')) {
                push(lines, 'default', JSON.stringify(pickString(param.default)));
              } else {
                push(lines, 'default', pickString(param.default));
              }
            }
            break;
          case 'multiline_string[]':
            if (config.target.includes('MV')) {
              push(lines, 'default', JSON.stringify(param.default.map((n) => JSON.stringify(pickString(n)))));
            } else {
              push(lines, 'default', JSON.stringify(param.default.map((n) => pickString(n))));
            }
            break;
          case 'note':
            if (param.default) push(lines, 'default', JSON.stringify(pickString(param.default)));
            break;
          case 'note[]':
            push(lines, 'default', JSON.stringify(param.default.map((n) => JSON.stringify(pickString(n)))));
            break;
          case 'number':
          case 'number[]':
            if (param.min !== undefined) push(lines, 'min', param.min.toString());
            if (param.max !== undefined) push(lines, 'max', param.max.toString());
            push(lines, 'decimals', param.decimals.toString());
            push(lines, 'default', JSON.stringify(param.default));
            break;
          case 'boolean':
          case 'boolean[]':
            push(lines, 'on', pickString(param.on));
            push(lines, 'off', pickString(param.off));
            push(lines, 'default', JSON.stringify(param.default));
            break;
          case 'file':
          case 'file[]':
            push(lines, 'dir', param.dir);
            if (param.type === 'file') {
              if (param.default) push(lines, 'default', param.default);
            } else {
              push(lines, 'default', JSON.stringify(param.default));
            }
            if (config.target.includes('MV')) push(lines, 'require', '1');
            break;
          case 'select':
          case 'select[]':
            param.options.forEach((option) => {
              push(lines, 'option', pickString(option.name));
              push(lines, 'value', option.value.toString());
            });
            if (param.type === 'select') push(lines, 'default', param.default);
            else push(lines, 'default', JSON.stringify(param.default));
            break;
          case 'combo':
          case 'combo[]':
            param.options.forEach((option) => {
              push(lines, 'option', pickString(option));
            });
            if (param.type === 'combo') push(lines, 'default', param.default);
            else push(lines, 'default', JSON.stringify(param.default));
            break;
          case 'actor':
          case 'class':
          case 'skill':
          case 'item':
          case 'weapon':
          case 'armor':
          case 'enemy':
          case 'troop':
          case 'state':
          case 'animation':
          case 'tileset':
          case 'common_event':
          case 'switch':
          case 'variable':
            push(lines, 'default', param.default.toString());
            break;
          case 'actor[]':
          case 'class[]':
          case 'skill[]':
          case 'item[]':
          case 'weapon[]':
          case 'armor[]':
          case 'enemy[]':
          case 'troop[]':
          case 'state[]':
          case 'animation[]':
          case 'tileset[]':
          case 'common_event[]':
          case 'switch[]':
          case 'variable[]':
            push(lines, 'default', JSON.stringify(param.default));
            break;
          case 'struct':
            push(lines, 'default', escapeStructParam(pickObject(param.default)));
            break;
          case 'struct[]': {
            if (Array.isArray(param.default)) {
              push(lines, 'default', JSON.stringify(param.default.map((n) => escapeStructParam(pickObject(n)))));
            } else if (param.default) {
              push(
                lines,
                'default',
                JSON.stringify(
                  pickObject(param.default).map((n: Record<string, any>) => escapeStructParam(pickObject(n))),
                ),
              );
            }
            break;
          }
          default: {
            const badParameter: never = param;
            throw new Error(`unknown parameter: ${badParameter}`);
          }
        }
      }

      function endSection(lines: string[]) {
        const len = lines.length;
        if (len === 0) return;
        if (lines[len - 1] === '') return;
        lines.push('');
      }

      function generateMain() {
        const lines: string[] = [];

        // base
        config.target.forEach((target) => push(lines, 'target', target));
        push(lines, 'plugindesc', pickString(config.title));
        if (config.version) push(lines, 'version', pickString(config.version));
        push(lines, 'author', pickString(config.author));
        if (config.license) push(lines, 'license', pickString(config.license));
        if (config.url) push(lines, 'url', pickString(config.url));
        endSection(lines);

        // dependencies
        config.base?.forEach((base) => push(lines, 'base', base));
        config.orderAfter?.forEach((orderAfter) => push(lines, 'orderAfter', orderAfter));
        config.orderBefore?.forEach((orderBefore) => push(lines, 'orderBefore', orderBefore));
        endSection(lines);

        // description
        if (config.help) push(lines, 'help', pickString(config.help));
        endSection(lines);

        // assets
        config.requiredAssets?.forEach((requiredAssets) => push(lines, 'requiredAssets', requiredAssets));
        endSection(lines);

        config.requiredNoteAssets?.forEach((requiredNoteAssets) => {
          push(lines, 'noteParam', requiredNoteAssets.name);
          push(lines, 'noteDir', requiredNoteAssets.dir);
          push(lines, 'noteType', requiredNoteAssets.type);
          push(lines, 'noteData', requiredNoteAssets.data);
          endSection(lines);
        });

        // parameter
        config.params?.forEach((param) => {
          pushParameter(lines, 'param', param);
          endSection(lines);
        });

        // commands
        config.commands?.forEach((command) => {
          push(lines, 'command', command.name);
          push(lines, 'text', pickString(command.text));
          push(lines, 'desc', pickString(command.description));
          endSection(lines);

          command.args.forEach((arg) => {
            pushParameter(lines, 'arg', arg);
            endSection(lines);
          });
        });

        if (lines[lines.length - 1] === '') lines.pop();
        return `/*:${omitLanguageSignature ? '' : language}\n * ${lines.join('\n * ')}\n */`;
      }

      function generateStruct() {
        return config.structs
          .map((struct) => {
            const lines: string[] = [];

            struct.params.forEach((param) => {
              pushParameter(lines, 'param', param);
              endSection(lines);
            });

            if (lines[lines.length - 1] === '') lines.pop();
            return `/*~struct~${struct.name}:${omitLanguageSignature ? '' : language}\n * ${lines.join('\n * ')}\n */`;
          })
          .join('\n\n');
      }

      return [generateMain(), generateStruct()].filter(Boolean).join('\n\n');
    })
    .join('\n\n')
    .split('\n')
    .map((line) => line.trimEnd())
    .join('\n');
}

function escapeStructParam(param: { [key: string]: any }) {
  const result: { [key: string]: string } = {};

  for (const key in param) {
    const value = param[key];
    if (Array.isArray(value)) {
      result[key] = JSON.stringify(value.map((n) => escapeStructParam(n)));
    } else if (typeof value === 'object') {
      result[key] = escapeStructParam(value);
    } else {
      result[key] = String(value);
    }
  }

  return JSON.stringify(result);
}
