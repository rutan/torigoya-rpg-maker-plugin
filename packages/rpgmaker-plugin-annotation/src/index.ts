export * from './generateAnnotation';
export * from './schema';
export * from './sanitize';
