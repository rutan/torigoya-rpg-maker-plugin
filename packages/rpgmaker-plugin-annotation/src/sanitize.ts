import { pluginConfigSchema } from './internal/schemaDefine';
import { PluginConfigSchema } from './schema';

export function sanitize(pluginConfig: PluginConfigSchema) {
  return pluginConfigSchema.parse(pluginConfig);
}
