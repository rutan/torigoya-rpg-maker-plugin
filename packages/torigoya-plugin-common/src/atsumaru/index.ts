export * from './comments';
export * from './getAtsumaru';
export * from './isAtsumaru';
