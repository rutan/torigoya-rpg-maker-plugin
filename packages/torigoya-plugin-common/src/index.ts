export * from './atsumaru/index';
export * from './utils/index';
export * from './getPluginName';
export * from './params';
export * from './Torigoya';
