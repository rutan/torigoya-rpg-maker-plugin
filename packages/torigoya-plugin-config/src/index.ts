export * from './types.js';
export * from './loadConfig.js';
export * from './writeAnnotation.js';
export * from './writeParameterReader.js';
export * from './rollup/defineRollupConfig.js';
export * from './rollup/pluginRPGMakerConfig.js';
