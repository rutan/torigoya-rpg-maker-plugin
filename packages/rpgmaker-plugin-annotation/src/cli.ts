#!/usr/bin/env node
import { program } from 'commander';
import * as fs from 'fs';
import { version } from '../package.json';
import { generateAnnotation } from './generateAnnotation';
import { sanitize } from './sanitize';

(() => {
  program.version(`v.${version}`);

  program
    .argument('<input>')
    .option('-o --out <out>', 'output file')
    .option('-l --languages <languages...>', 'languages')
    .option('--defaultLanguage <defaultLanguage>', 'default language')
    .showHelpAfterError()
    .action((input, options, _command) => {
      const languages = options.languages || ['en', 'ja'];
      const defaultLanguage = options.defaultLanguage || languages[0];

      const json = fs.readFileSync(input, 'utf-8');
      const data = JSON.parse(json);
      const result = generateAnnotation(sanitize(data), { languages, defaultLanguage });

      if (options.out) {
        fs.writeFileSync(options.out, result, 'utf-8');
      } else {
        console.log(result);
      }
    });

  program.parse(process.argv);
})();
