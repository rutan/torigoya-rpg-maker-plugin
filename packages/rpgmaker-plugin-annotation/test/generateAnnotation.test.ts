import { generateAnnotation, sanitize } from '../src';
import * as sample from './fixture/sample.json';

test('generateAnnotation snapshot', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const config = sanitize(sample as any);

  const result = generateAnnotation(config, {
    languages: ['en', 'ja'],
    defaultLanguage: 'en',
  });
  expect(result).toMatchSnapshot();
});
