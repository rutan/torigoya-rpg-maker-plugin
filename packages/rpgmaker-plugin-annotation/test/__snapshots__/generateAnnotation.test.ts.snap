// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAnnotation snapshot 1`] = `
"/*:
 * @target MZ
 * @plugindesc SamplePlugin
 * @version 1.2.3
 * @author Name
 * @license Public domain
 * @url https://example.com
 *
 * @base ParentPlugin
 * @base ParentPluginEx
 * @orderAfter ChildPlugin
 * @orderAfter ChildPluginEx
 * @orderBefore ParentPlugin2
 * @orderBefore ParentPlugin2Ex
 *
 * @help This is a sample plugin.
 * Good luck :)
 *
 * @requiredAssets img/pictures/image_1
 * @requiredAssets img/system/image_2
 *
 * @noteParam name
 * @noteDir img/pictures
 * @noteType file
 * @noteData actors
 *
 * @param stringParameter
 * @text String Parameter
 * @desc this is a string
 * @type string
 * @default hello, world
 *
 * @param stringParameterWithEmptyDefault
 * @text String Parameter with Empty Default Value
 * @desc this is a string
 * @type string
 *
 * @param stringArrayParameter
 * @text String Array Parameter
 * @desc this is a string array
 * @type string[]
 * @default ["hello","world"]
 *
 * @param multilineStringParameter
 * @text MultilineString Parameter
 * @desc this is a multiline string
 * @type multiline_string
 * @default hello
 * world
 *
 * @param multilineStringArrayParameter
 * @text MultilineString Array Parameter
 * @desc this is a multiline string array
 * @type multiline_string[]
 * @default ["hello\\nworld","happy\\nworld"]
 *
 * @param noteParameter
 * @text Note Parameter
 * @desc this is a note
 * @type note
 * @default "<Hello:World>"
 *
 * @param noteArrayParameter
 * @text Note Array Parameter
 * @desc this is a note array
 * @type note[]
 * @default ["\\"<Hello:World>\\"","\\"<Happy:World>\\""]
 *
 * @param numberParameter
 * @text Number Parameter
 * @desc this is a number
 * @type number
 * @min 100
 * @max 100000
 * @decimals 0
 * @default 0
 *
 * @param numberParameter
 * @text Number Array Parameter
 * @desc this is a number array
 * @type number[]
 * @min -100
 * @max 1000
 * @decimals 2
 * @default [123.45,456.67]
 *
 * @param booleanParameter
 * @text Boolean Parameter
 * @desc this is a boolean
 * @type boolean
 * @on ON
 * @off OFF
 * @default false
 *
 * @param booleanParameter
 * @text Boolean Array Parameter
 * @desc this is a boolean array
 * @type boolean[]
 * @on TRUE
 * @off FALSE
 * @default [false,true]
 *
 * @param fileParameter
 * @text File Parameter
 * @desc this is a file
 * @type file
 * @dir img/pictures
 * @default Actor_01
 *
 * @param fileArrayParameter
 * @text File Array Parameter
 * @desc this is a file array
 * @type file[]
 * @dir img/pictures
 * @default ["Actor_01","Actor_02"]
 *
 * @param actorParameter
 * @text Actor Parameter
 * @desc this is a actor
 * @type actor
 * @default 1
 *
 * @param actorArrayParameter
 * @text Actor Array Parameter
 * @desc this is a actor array
 * @type actor[]
 * @default [1,2,3,4]
 *
 * @param selectParameter
 * @text Select Parameter
 * @desc this is a select parameter
 * @type select
 * @option Option 1
 * @value A
 * @option Option-2
 * @value B
 * @default A
 *
 * @param selectArrayParameter
 * @text Select Array Parameter
 * @desc this is a select parameter array
 * @type select[]
 * @option Option 1
 * @value A
 * @option Option-2
 * @value B
 * @default ["A","B"]
 *
 * @param structParameter
 * @text Struct Parameter
 * @desc this is a struct parameter
 * @type struct<MyStructParent>
 * @default {"structParam1":"10","structParam2":"This is a string!","structParam3":"{\\"structParam1\\":\\"20\\"}"}
 *
 * @param structParameterArray1
 * @text Struct Parameter Array
 * @desc this is struct parameters
 * @type struct<MyStructParent>[]
 * @default ["{\\"structParam1\\":\\"100\\",\\"structParam2\\":\\"This is a string! / これは文字列です！\\",\\"structParam3\\":\\"{\\\\\\"structParam1\\\\\\":\\\\\\"200\\\\\\"}\\"}"]
 *
 * @param structParameterArray2
 * @text Struct Parameter Array (locale)
 * @desc this is struct parameters
 * @type struct<MyStructParent>[]
 * @default ["{\\"structParam1\\":\\"123\\",\\"structParam2\\":\\"This is a english string!\\",\\"structParam3\\":\\"{\\\\\\"structParam1\\\\\\":\\\\\\"234\\\\\\"}\\"}"]
 *
 * @command MyCommand
 * @text My Original Command
 * @desc This is a super command.
 * This command is very useful.
 *
 * @arg commandParam1
 * @text Command Parameter 1
 * @desc This is a boolean
 * @type boolean
 * @on TRUE
 * @off FALSE
 * @default false
 */

/*~struct~MyStructParent:
 * @param structParam1
 * @text MyStruct Parameter 1
 * @desc This is a number
 * @type number
 * @decimals 0
 * @default 12345
 *
 * @param structParam2
 * @text MyStruct Parameter 2
 * @desc This is a string
 * @type string
 * @default this is a string
 *
 * @param structParam3
 * @text MyStruct Parameter 3
 * @desc This is a struct
 * @type struct<MyStructChild>
 * @default {"structParam1":"20"}
 *
 * @param structParam4
 * @text MyStruct Parameter 4
 * @desc This is a struct array
 * @type struct<MyStructChild>[]
 * @default ["{\\"structParam1\\":\\"1\\"}","{\\"structParam1\\":\\"2\\"}","{\\"structParam1\\":\\"3\\"}"]
 */

/*~struct~MyStructChild:
 * @param structParam1
 * @text MyStruct Parameter 1
 * @desc This is a number
 * @type number
 * @decimals 0
 * @default 12345
 */

/*:ja
 * @target MZ
 * @plugindesc サンプルプラグイン
 * @version 1.2.3
 * @author Name
 * @license Public domain
 * @url https://example.com
 *
 * @base ParentPlugin
 * @base ParentPluginEx
 * @orderAfter ChildPlugin
 * @orderAfter ChildPluginEx
 * @orderBefore ParentPlugin2
 * @orderBefore ParentPlugin2Ex
 *
 * @help This is a sample plugin.
 * Good luck :)
 *
 * @requiredAssets img/pictures/image_1
 * @requiredAssets img/system/image_2
 *
 * @noteParam name
 * @noteDir img/pictures
 * @noteType file
 * @noteData actors
 *
 * @param stringParameter
 * @text 文字列パラメータ
 * @desc this is a string
 * @type string
 * @default こんにちは、世界
 *
 * @param stringParameterWithEmptyDefault
 * @text 文字列パラメータ（デフォルト値空欄）
 * @desc this is a string
 * @type string
 *
 * @param stringArrayParameter
 * @text 文字列配列パラメータ
 * @desc this is a string array
 * @type string[]
 * @default ["hello","world"]
 *
 * @param multilineStringParameter
 * @text 複数行文字列パラメータ
 * @desc this is a multiline string
 * @type multiline_string
 * @default hello
 * world
 *
 * @param multilineStringArrayParameter
 * @text 複数行文字列配列パラメータ
 * @desc this is a multiline string array
 * @type multiline_string[]
 * @default ["hello\\nworld","happy\\nworld"]
 *
 * @param noteParameter
 * @text メモ欄パラメータ
 * @desc this is a note
 * @type note
 * @default "<Hello:World>"
 *
 * @param noteArrayParameter
 * @text メモ欄配列パラメータ
 * @desc this is a note array
 * @type note[]
 * @default ["\\"<Hello:World>\\"","\\"<Happy:World>\\""]
 *
 * @param numberParameter
 * @text 数値パラメータ
 * @desc this is a number
 * @type number
 * @min 100
 * @max 100000
 * @decimals 0
 * @default 0
 *
 * @param numberParameter
 * @text 数値配列パラメータ
 * @desc this is a number array
 * @type number[]
 * @min -100
 * @max 1000
 * @decimals 2
 * @default [123.45,456.67]
 *
 * @param booleanParameter
 * @text 真偽値パラメータ
 * @desc this is a boolean
 * @type boolean
 * @on オン
 * @off オフ
 * @default false
 *
 * @param booleanParameter
 * @text 真偽値配列パラメータ
 * @desc this is a boolean array
 * @type boolean[]
 * @on TRUE
 * @off FALSE
 * @default [false,true]
 *
 * @param fileParameter
 * @text ファイルパラメータ
 * @desc this is a file
 * @type file
 * @dir img/pictures
 * @default Actor_01
 *
 * @param fileArrayParameter
 * @text ファイル配列パラメータ
 * @desc this is a file array
 * @type file[]
 * @dir img/pictures
 * @default ["Actor_01","Actor_02"]
 *
 * @param actorParameter
 * @text アクターパラメータ
 * @desc this is a actor
 * @type actor
 * @default 1
 *
 * @param actorArrayParameter
 * @text アクター配列パラメータ
 * @desc this is a actor array
 * @type actor[]
 * @default [1,2,3,4]
 *
 * @param selectParameter
 * @text セレクトパラメータ
 * @desc this is a select parameter
 * @type select
 * @option オプション1
 * @value A
 * @option Option-2
 * @value B
 * @default A
 *
 * @param selectArrayParameter
 * @text セレクト配列パラメータ
 * @desc this is a select parameter array
 * @type select[]
 * @option オプション1
 * @value A
 * @option Option-2
 * @value B
 * @default ["A","B"]
 *
 * @param structParameter
 * @text 構造体パラメータ
 * @desc this is a struct parameter
 * @type struct<MyStructParent>
 * @default {"structParam1":"10","structParam2":"これは文字列だ！","structParam3":"{\\"structParam1\\":\\"20\\"}"}
 *
 * @param structParameterArray1
 * @text 構造体パラメータ配列
 * @desc this is struct parameters
 * @type struct<MyStructParent>[]
 * @default ["{\\"structParam1\\":\\"100\\",\\"structParam2\\":\\"This is a string! / これは文字列です！\\",\\"structParam3\\":\\"{\\\\\\"structParam1\\\\\\":\\\\\\"200\\\\\\"}\\"}"]
 *
 * @param structParameterArray2
 * @text 構造体パラメータ配列（言語設定）
 * @desc this is struct parameters
 * @type struct<MyStructParent>[]
 * @default ["{\\"structParam1\\":\\"321\\",\\"structParam2\\":\\"これは日本語の文字列です！\\",\\"structParam3\\":\\"{\\\\\\"structParam1\\\\\\":\\\\\\"432\\\\\\"}\\"}"]
 *
 * @command MyCommand
 * @text My Original Command
 * @desc This is a super command.
 * This command is very useful.
 *
 * @arg commandParam1
 * @text Command Parameter 1
 * @desc This is a boolean
 * @type boolean
 * @on TRUE
 * @off FALSE
 * @default false
 */

/*~struct~MyStructParent:ja
 * @param structParam1
 * @text MyStruct Parameter 1
 * @desc This is a number
 * @type number
 * @decimals 0
 * @default 12345
 *
 * @param structParam2
 * @text MyStruct Parameter 2
 * @desc This is a string
 * @type string
 * @default これは文字列
 *
 * @param structParam3
 * @text MyStruct Parameter 3
 * @desc This is a struct
 * @type struct<MyStructChild>
 * @default {"structParam1":"20"}
 *
 * @param structParam4
 * @text MyStruct Parameter 4
 * @desc This is a struct array
 * @type struct<MyStructChild>[]
 * @default ["{\\"structParam1\\":\\"1\\"}","{\\"structParam1\\":\\"2\\"}","{\\"structParam1\\":\\"3\\"}"]
 */

/*~struct~MyStructChild:ja
 * @param structParam1
 * @text MyStruct Parameter 1
 * @desc This is a number
 * @type number
 * @decimals 0
 * @default 12345
 */"
`;
